py/dml/structure.py

```diff
@@ -1414,51 +1414,95 @@ def wrap_method_body_in_try(site, overridden_site, obj, name, body,
+    shared = True
+    def __init__(self, name, vtable_trait):
+        (site, *sig) = vtable_trait.vtable_methods[name]
+        obj_spec = dml.globals.templates[vtable_trait.name].spec
+        super(AbstractTraitMethodHandle, self).__init__(
+            site, name, obj_spec, True, True, None, *sig)
+
+
+
+def process_method_declarations(obj, name, declarations,
+                                shared_impl_traits, shared_absdecl_traits,
+                                obj_specs, vtable_nothrow_dml14):
+    # A method can have both shared and non-shared declarations.
     # We will create a MethodHandle object for either, which
-    # sort_method_implementations() uses to resolve override order.
+    # sort_method_declarations() uses to resolve override order.
     unshared_methods = [
         ObjMethodHandle(method_ast, obj_spec)
-        for (obj_spec, method_ast) in implementations]
+        for (obj_spec, method_ast) in declarations]
     shared_methods = [
         TraitMethodHandle(
             impl_trait.method_impls[name],
             dml.globals.templates[impl_trait.name].spec)
         for impl_trait in shared_impl_traits]
+    shared_abstract_methods = [
+        AbstractTraitMethodHandle(name, vtable_trait)
+        for vtable_trait in shared_absdecl_traits]
 
-    (default_map, method_order) = sort_method_implementations(
-        unshared_methods + shared_methods, obj_specs)
+    (default_map, decl_ancestry_map, method_order,
+     abstract_decls) = sort_method_declarations(
+        unshared_methods + shared_methods + shared_abstract_methods, obj_specs)
 
     location = Location(obj, static_indices(obj))
 
-    impl_to_method = {}
-    for (default_level, impl) in reversed(list(enumerate(
-            method_order))):
+    nonshared_impls = []
+
+    for impl in method_order:
         if impl.shared:
             if default_map[impl]:
                 # shared method overrides a non-shared method
                 report(ETMETH(
                     default_map[impl][0].site, impl.site, name))
-            # handled separately
-            continue
+        else:
+            nonshared_impls.append(impl)
+
+    if not nonshared_impls:
+        if shared_methods:
+            shared_impl_sig = shared_methods[0].signature
+            for decl in abstract_decls:
+                if decl.shared:
+                    # handled separately
+                    continue
+
+                try:
+                    traits.typecheck_method_override(shared_impl_sig,
+                                                     decl.signature)
+                except DMLError as e:
+                    report(e)
+            return None
+        else:
+            assert abstract_decls
+            for decl in abstract_decls:
+                # We favor reporting EABSTEMPLATE (done by the caller)
+                # over EABSMETH
+                if not decl.shared:
+                    report(EABSMETH(decl.site, name))
```

Comment:
It would maybe be cleaner if checking abstract methods+params was a single responsibility? The overlap between EABSMETH and EABSTEMPLATE feels somewhat unnatural, maybe it would suffice with EABSMETH and ENPARAM, and give both an optional template arg? or even a mandatory template arg, that may be a file?

In any case: no need to fix within this PR.